### PR TITLE
Fixes a bug that causes `CSVLogger` to overwrite `version_0` when `root_dir` is a relative path.

### LINKED ISSUE #17138: CSVLogger overwriting previous logs when giving relative root dir
### Bug description

When I upgraded to PL2.0, I found the `CSVLogger` repeatedly writes to `version_0`, when the `root_dir` is given a partial path like `CSVLogger('my_csv_logs')`. A closer look reveals that the `_get_next_version` implementation in `lightning/fabric/loggers/csv_logs.py` is problematic. Interestingly, there isn't a similar issue for tensorboard loggers. I have submitted a PR since the fix is simple (https://github.com/Lightning-AI/lightning/pull/17139). 

PS: the `_get_next_version` function is implemented twice in the fabric loggers (`lightning/fabric/loggers/tensorboard.py`) and the lightning/pytorch loggers (`lightning/pytorch/loggers/tensonboard.py`), making the former code useless). This issue is not present for `CSVLoggers`, though, and may be of your interest when refactoring the code.

### How to reproduce the bug

```python
# use any model
csv_logger = CSVLogger('my_csv_logs')
pl.Trainer(logger=csv_logger)
```

### Error messages and logs

```
# Error messages and logs here please
/home/[redacted]/miniconda3/envs/py310/lib/python3.10/site-packages/pytorch_lightning/callbacks/model_checkpoint.py:612: UserWarning: Checkpoint directory [redacted]/lightning_logs/version_0/checkpoints exists and is not empty.
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0): 2.0.0
#- Lightning App Version (e.g., 0.5.2): ?
#- PyTorch Version (e.g., 2.0): 2.0.0
#- Python version (e.g., 3.9): 3.10
#- OS (e.g., Linux): Ubuntu Linux 20.04
#- CUDA/cuDNN version: 11.8
#- GPU models and configuration: RTX 3090
#- How you installed Lightning(`conda`, `pip`, source): pip
#- Running environment of LightningApp (e.g. local, cloud): ?
```

</details>

### More info

_No response_

cc @borda

### PR DESCRIPTION
**Pull request by water-vapor**
## What does this PR do?

Fixes a bug that causes `CSVLogger` to overwrite `version_0` when `root_dir` is a relative path.

Fixes #17138

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/fabric/loggers/csv_logs.py [hunk 1]
```diff
@@ -158,9 +158,10 @@ def _get_next_version(self) -> int:
             return 0
 
         existing_versions = []
-        for d in self._fs.listdir(root_dir, detail=False):
-            name = d[len(root_dir) + 1 :]  # removes parent directories
-            if self._fs.isdir(d) and name.startswith("version_"):
+        for d in self._fs.listdir(root_dir):
```
Comment (Borda) [inline src/lightning/fabric/loggers/csv_logs.py:161]: lets add a test for this

Comment (water-vapor) [reply src/lightning/fabric/loggers/csv_logs.py:161]: I added a test for this case. And I can confirm that the test failed before, and passed after this fix.

**Review by Borda — APPROVED** [review]

File: tests/tests_fabric/loggers/test_csv.py [hunk 2]
```diff
@@ -30,6 +30,16 @@ def test_file_logger_automatic_versioning(tmpdir):
     assert logger.version == 2
 
 
+def test_file_logger_automatic_versioning_relative_root_dir(tmpdir):
+    """Verify that automatic versioning works, when root_dir is given a relative path."""
+    root_dir = tmpdir.mkdir("exp")
+    root_dir.mkdir("version_0")
+    root_dir.mkdir("version_1")
+    os.chdir(tmpdir)
+    logger = CSVLogger(root_dir="exp", name="exp")
```
Comment (awaelchli) [inline tests/tests_fabric/loggers/test_csv.py:39]:
```suggestion
    logger = CSVLogger(root_dir="exp", name="name")
```

**Comment by awaelchli** [comment]
This was probably introduced here: #16880
So I'm setting the milestone to 2.0.x for cherry-picking.

File: tests/tests_fabric/loggers/test_csv.py (hunk 2 above)
Comment (water-vapor) [reply tests/tests_fabric/loggers/test_csv.py:39]: @awaelchli  Suppose the logging path looks like `some_dir/exp/version_0`, this `root_dir` should be `some_dir/exp` according to its internal usage (more info at https://github.com/Lightning-AI/lightning/issues/17155). Though the name parameter is not used internally by the CSVLogger to calculate the next version, from my understanding it should refer to the `exp` directory. May I ask why we'd want to use `name="name"` instead?

Comment (water-vapor) [reply tests/tests_fabric/loggers/test_csv.py:39]: @awaelchli My point is that the `root_dir` should be equal to the `name` parameter because they refer to the same directory. If they have to be different, you can create a nested directory for testing, like `root_dir=dir1/dir2` and `name=dir2`.

Comment (awaelchli) [reply tests/tests_fabric/loggers/test_csv.py]: root_dir and name cannot refer to the same directory, that's not allowed. We need to guarantee the following folder structure: `root_dir/name/version`. 

So if you do `CSVLogger(root_dir="exp", name="exp")` we expect this `./exp/exp/version` to be the path where logs get saved.

Comment (awaelchli) [reply tests/tests_fabric/loggers/test_csv.py]: In your other example `CSVLogger(root_dir="dir1/dir2", name="dir2")` would result in a directory structure of `./dir1/dir2/dir2` which is expected.

Comment (water-vapor) [reply tests/tests_fabric/loggers/test_csv.py]: @awaelchli Yes, but unfortunately, that is the current behavior for pytorch.loggers.CSVLogger, but not fabric CSVLogger, and that issue is what I described in https://github.com/Lightning-AI/lightning/issues/17155.

Comment (water-vapor) [reply tests/tests_fabric/loggers/test_csv.py]: @awaelchli This issue has another fix, which will produce the correct behavior you described, and was my first attempt. The fix will add `save_dir=os.path.join(self.root_dir, self.name)` right after the first line of `_get_next_version`. But this will make all other tests about versions in `test_csv.py` fail, since other tests are passing `root_dir` the same as `name`

Comment (water-vapor) [reply tests/tests_fabric/loggers/test_csv.py]: @awaelchli Is it possible to get this bug fixed in whatever way you prefer? The bug itself is simple to spot and fix and is not entangled with any refactoring or design issues.

**Review by lantiga — APPROVED** [review]
LGTM

**Review by justusschock — APPROVED** [review]
